sklego/preprocessing/intervalencoder.py

```diff
@@ -0,0 +1,144 @@
+    becomes the value of sigma that is used for weighted averaging
+    :return:
+        An array as long as `intervals` that represents the average `y`-values at those intervals.
+    """
+    results = np.zeros(intervals.shape)
+    for idx, interval in enumerate(intervals):
+        if method == "average":
+            distances = 1 / (0.01 + np.abs(xs - interval))
+            predicate = (xs < (interval + span)) | (xs < (interval - span))
+        elif method == "normal":
+            distances = np.exp(-((xs - interval) ** 2) / span)
+            predicate = xs == xs
+        else:
+            raise ValueError("method needs to be either `average` or `normal`")
+        subset = ys[predicate]
+        dist_subset = distances[predicate]
+        results[idx] = np.average(subset, weights=dist_subset)
+    return results
+
+
+class IntervalEncoder(TransformerMixin, BaseEstimator):
+    """
+    The interval encoder bends features in `X` with regards to`y`.
+    We take each column in X separately and smooth it towards `y` using
+    the strategy that is defined in `method`.
+    Note that this allows us to make certain features strictly monotonic
+    in your machine learning model if you follow this with an appropriate
+    model.
+    :param n_chunks: the number of cuts that makes the interval
+    :param method: the interpolation method used, must be in ["normal", "average"], default: "normal"
+    :param span: a hyperparameter for the interpolation method, if the
+    method is `normal` it resembles the width of the radial basis
+    function used to weigh the points
+    """
+
+    def __init__(self, n_chunks=10, span=1, method="normal"):
+        self.span = span
+        self.method = method
+        self.n_chunks = n_chunks
+
+    def fit(self, X, y):
+        """Fits the estimator"""
+        allowed_methods = ["average", "normal", "increasing", "decreasing"]
+        if self.method not in allowed_methods:
+            raise ValueError(
+                f"`method` must be in {allowed_methods}, got `{self.method}`"
+            )
+        if self.n_chunks <= 0:
+            raise ValueError(f"`n_chunks` must be >= 1, received {self.n_chunks}")
+        if self.span > 1.0:
+            raise ValueError(
+                f"Error, we expect 0 <= span <= 1, received span={self.span}"
+            )
+        if self.span < 0.0:
+            raise ValueError(
+                f"Error, we expect 0 <= span <= 1, received span={self.span}"
+            )
+
+        # these two matrices will have shape (columns, quantiles)
+        # quantiles indicate where the interval split occurs
+        X, y = check_X_y(X, y, estimator=self)
+        self.quantiles_ = np.zeros((X.shape[1], self.n_chunks))
+        # heights indicate what heights these intervals will have
+        self.heights_ = np.zeros((X.shape[1], self.n_chunks))
+        self.num_cols_ = X.shape[1]
+
+        average_func = (
+            _mk_average
```

Comment:
The dispatching here feels a bit double. First we're dispatching to `_mk_average` and `_mk_monotonic_average`, but inside those methods we're dispatching again based on `method`. How about something like this:

```
average_func = {
   'normal': _mk_normal,
   'average': _mk_average,
   'increasing': partial(_mk_monotonic, constraint_cmp=operator.ge),
   'decreasing':  partial(_mk_monotonic,  constraint_cmp=operator.le),
}[self.method]

Reply:
i was wondering about this. the downside of having two functions `_mk_normal` and `_mk_average` is that you'll end up with two functions with *a lot* or repitition, doing against the DRY principle. 

by having the code run this way you'll get `*_average` functions that all accept a `method` param that feels somewhat general.

Reply:
maybe one other thing; there is going to be one extra function added for the `updown` and `downup` relation. it might be preferable to delay any refactoring until it is clear how that function is supposed to work.

Reply:
Yep, seems like a good idea!